astropy/nddata/utils.py

```diff
@@ -456,3 +469,322 @@ def block_replicate(data, block_size, conserve_sum=True):
+
+        shape : tuple
+            The shape (``(ny, nx)``) of the cutout array in pixel
+            coordinates (but see the ``mode`` keyword for additional
+            details).
+
+        wcs : `~astropy.wcs.WCS`, optional
+            A WCS object associated with the input ``data`` array.  If
+            ``wcs`` is not `None`, then the returned cutout object will
+            contain a copy of the updated WCS for the cutout data array.
+
+        mode : {'trim', 'partial', 'strict'}, optional
+            The mode used for creating the cutout data array.  For the
+            ``'partial'`` and ``'trim'`` modes, a partial overlap of the
+            cutout array and the input ``data`` array is sufficient.
+            For the ``'strict'`` mode, the cutout array has to be fully
+            contained within the ``data`` array, otherwise an
+            `~astropy.nddata.utils.PartialOverlapError` is raised.   In
+            all modes, non-overlapping arrays will raise a
+            `~astropy.nddata.utils.NoOverlapError`.  In ``'partial'``
+            mode, positions in the cutout array that do not overlap with
+            the ``data`` array will be filled with ``fill_value``.  In
+            ``'trim'`` mode only the overlapping elements are returned,
+            thus the resulting cutout array may be smaller than the
+            requested ``shape``.
+
+        fill_value : number, optional
+            If ``mode='partial'``, the value to fill pixels in the
+            cutout array that do not overlap with the input ``data``.
+            ``fill_value`` must have the same ``dtype`` as the input
+            ``data`` array.
+
+        copy : bool, optional
+            If `False` (default), then the cutout data will be a view
+            into the original ``data`` array.  If `True`, then the
+            cutout data will hold a copy of the original ``data`` array.
+
+        Returns
+        -------
+        result : `~astropy.nddata.utils.Cutout2D`
+            A cutout object containing the 2D cutout data array and the
+            updated WCS, if ``wcs`` is input.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from astropy.nddata.utils import Cutout2D
+        >>> data = np.arange(20.).reshape(5, 4)
+        >>> c1 = Cutout2D(data, (2, 2), (3, 3))
+        >>> print(c1.data)
+        [[  5.   6.   7.]
+         [  9.  10.  11.]
+         [ 13.  14.  15.]]
+
+        >>> print(c1.center_original)
+        (2.0, 2.0)
+        >>> print(c1.center_cutout)
+        (1.0, 1.0)
+        >>> print(c1.origin_original)
+        (1, 1)
+
+        >>> c2 = Cutout2D(data, (0, 0), (3, 3))
+        >>> print(c2.data)
+        [[ 0.  1.]
+         [ 4.  5.]]
+
+        >>> c3 = Cutout2D(data, (0, 0), (3, 3), mode='partial')
+        >>> print(c3.data)
+        [[ nan  nan  nan]
+         [ nan   0.   1.]
+         [ nan   4.   5.]]
+        """
+
+        if mode not in ['partial', 'trim', 'strict']:
```

Comment:
Yes, it is caught now in both `overlap_slices` and `extract_array`.  `extract_array` orginally didn't have the `'trim'` option (I added it in this PR) when I wrote that check.  I removed the now redundant `mode` check in `Cutout2D`.  Thanks for catching that.